bubuku/features/rebalance/change.py

```diff
@@ -0,0 +1,248 @@
+    if not items:
+        return
+    items = sorted(items, key=weight_key, reverse=True)
+    ceil_amount = int(amount / len(items))
+    amounts = [ceil_amount for _ in range(0, len(items))]
+    for idx in range(0, min(amount - sum(amounts), len(items))):
+        amounts[idx] += 1
+    for idx in range(0, len(items)):
+        apply_func(items[idx], amounts[idx])
+
+
+class OptimizedRebalanceChange(BaseRebalanceChange):
+    _LOAD_STATE = 'load_state'
+    _COMPUTE_LEADERS = 'compute_leaders'
+    _COMPUTE_REPLICAS = 'compute_replicas'
+    _SORT_ACTIONS = 'sort_actions'
+    _BALANCE = 'balance'
+
+    def __init__(self, zk: BukuExhibitor, broker_ids: list):
+        self.zk = zk
+        self.broker_ids = sorted(int(id_) for id_ in broker_ids)
+        self.broker_distribution = None
+        self.source_distribution = None
+        self.action_queue = None
+        self.state = OptimizedRebalanceChange._LOAD_STATE
+
+    def __str__(self):
+        return 'OptimizedRebalance state={}, queue_size={}'.format(
+            self.state, len(self.action_queue) if self.action_queue is not None else None)
+
+    def run(self, current_actions) -> bool:
+        # Stop rebalance if someone is restarting
+        if self.should_be_paused(current_actions):
+            _LOG.warning("Rebalance paused, because other blocking events running: {}".format(current_actions))
+            return True
+        if self.zk.is_rebalancing():
+            return True
+        new_broker_ids = sorted(int(id_) for id_ in self.zk.get_broker_ids())
+        if new_broker_ids != self.broker_ids:
+            _LOG.warning("Rebalance stopped because of broker list change from {} to {}".format(
+                self.broker_ids, new_broker_ids))
+            return False
+        if self.state == OptimizedRebalanceChange._LOAD_STATE:
+            self._load_data()
+            self.state = OptimizedRebalanceChange._COMPUTE_LEADERS
+        elif self.state == OptimizedRebalanceChange._COMPUTE_LEADERS:
+            self.action_queue = self.rebalance_leaders(self.broker_distribution)
+            self.state = OptimizedRebalanceChange._COMPUTE_REPLICAS
+        elif self.state == OptimizedRebalanceChange._COMPUTE_REPLICAS:
+            self._rebalance_replicas()
+            self.state = OptimizedRebalanceChange._SORT_ACTIONS
+        elif self.state == OptimizedRebalanceChange._SORT_ACTIONS:
+            self.action_queue = self._sort_actions()
+            self.state = OptimizedRebalanceChange._BALANCE
+        elif self.state == OptimizedRebalanceChange._BALANCE:
+            return not self._balance()
+        return True
+
+    def _balance(self):
+        if not self.action_queue:
+            return True
+        key, replicas = self.action_queue.popitem()
+        topic, partition = key
+        if not self.zk.reallocate_partition(topic, partition, replicas):
+            self.action_queue[key] = replicas
+        return False
```

Comment:
well, return not self._balance() for me return true if not balanced, false - if balanced.